### PR TITLE
PA-1615:Service Identity Automation - Automate Test cases-Initial Commit

### PR DESCRIPTION
**Pull request by jkagliwal-px**
Initial Commit for Service Identity feature automation.

Added supporting libraries and added the Auth-token context switch feature.
Added the first testcase as well, which performs the below steps - 

 1. Create Service Identity with PDS Admin User 
 2. IAM roles - for service id a. namespace-admin - ns1 b. namespace-reader - ns2 
 3. Deploy TLS Enabled PG on ns1 with service identity token 
 4. Validate the Deployment 
 5. Run Workloads 
 6. Take Backup
 7. Restore to ns2 -> expect failure 
 8. Update IAM for ns2 with admin role
 9. Restore again on ns2 -> restore shuould be successful
 10. Scale up restored dep on ns2
 11. Clear all created entities.

Bug reported on step 6 -> https://portworx.atlassian.net/browse/DS-6645

Aetos log link - https://aetos.pwx.purestorage.com/resultSet/testSetID/354899

### CODE REVIEW

File: drivers/pds/api/api_helper.go [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+	apiConf := pds.NewConfiguration()
+	apiConf.Host = endpointURL.Host
+	apiConf.Scheme = endpointURL.Scheme
+	if serviceIdFlag == true && strings.Contains(testName, "ServiceIdentity") {
+		serviceIdToken := siID.ReturnServiceIdToken()
+		if serviceIdToken == "" {
+			token, err = getBearerToken()
+		} else {
+			token = serviceIdToken
+			log.InfoD("***** Token with ServiceIdentity is %v-", serviceIdToken)
```
Comment (Madanagopal19) [inline drivers/pds/api/api_helper.go:68]: nit: remove ***

File: drivers/pds/api/api_helper.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+func GetContext() (context.Context, error) {
+	var token string
+	currentTestDescription := ginkgo.CurrentGinkgoTestDescription()
+	testName := strings.Split(currentTestDescription.FullTestText, " ")[0]
+	serviceIdFlag := customParams.ReturnServiceIdentityFlag()
+	log.InfoD("params.InfraToTest.ServiceIdentityToken is - %v", serviceIdFlag)
+	PDSControlPlaneURL := os.Getenv("CONTROL_PLANE_URL")
+	endpointURL, err := url.Parse(PDSControlPlaneURL)
+	if err != nil {
+		log.Errorf("Unable to access the URL: %s", PDSControlPlaneURL)
```
Comment (Madanagopal19) [inline drivers/pds/api/api_helper.go:56]: do not log the error

File: drivers/pds/parameters/parameters_utils.go [hunk 3]
```diff
@@ -111,3 +115,33 @@ func (customparams *Customparams) ReadParams(filename string) (*Parameter, error
 	}
 	return &jsonPara, nil
 }
+
+func (customparams *Customparams) UpdatePdsParams(params *Parameter) {
+	json.Marshal(params)
+	return
+}
+
+func (customparams *Customparams) SetServiceIdentityFlag() {
```
Comment (Madanagopal19) [inline drivers/pds/parameters/parameters_utils.go:124]: do we require this flag to be set ?

File: drivers/pds/pdsutils/utils.go [hunk 4]
```diff
@@ -1,88 +1,88 @@
...
-	var bearerToken = new(BearerToken)
-	err = json.Unmarshal(body, &bearerToken)
-	if err != nil {
-		return "", err
-	}
-
-	return bearerToken.AccessToken, nil
-
-}
+//import (
```
Comment (Madanagopal19) [inline drivers/pds/pdsutils/utils.go:3]: remove commented lines

File: tests/pds/pds_ds_rbac_test.go [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+			Timeout:        params.LoadGen.Timeout,
+			Replicas:       params.LoadGen.Replicas,
+			TableName:      params.LoadGen.TableName,
+			Iterations:     params.LoadGen.Iterations,
+			FailOnError:    params.LoadGen.FailOnError,
+		}
+	})
+
+	It("Deploy Dataservices", func() {
+		var deploymentsToBeCleaned []*pds.ModelsDeployment
```
Comment (Madanagopal19) [inline tests/pds/pds_ds_rbac_test.go:42]: nit : use var ()

File: tests/pds/pds_ds_rbac_test.go [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+			}
+
+		})
+		JustAfterEach(func() {
+			defer EndTorpedoTest()
+		})
+	})
+})
+
+func DeployandValidateDataServicesWithSiAndTls(ds dataservice.PDSDataService, namespaceName string, namespaceid, projectID string, resourceTemplateID string, appConfigID string, dsVersion string, dsImage string, dsID string) (*pds.ModelsDeployment, map[string][]string, map[string][]string, error) {
```
Comment (Madanagopal19) [inline tests/pds/pds_ds_rbac_test.go:230]: move this func to pds_common.go

**Review by Madanagopal19 — CHANGES_REQUESTED** [review]

File: drivers/pds/pdsutils/utils.go (hunk 4 above)
Comment (jkagliwal-px) [reply drivers/pds/pdsutils/utils.go:3]: Deleted the file, no longer in use.

File: tests/pds/pds_ds_rbac_test.go (hunk 6 above)
Comment (jkagliwal-px) [reply tests/pds/pds_ds_rbac_test.go:230]: Cannot move it, as unable to access few variables like "storagetemplateid" and few other vars.

File: drivers/pds/api/api_helper.go [hunk 7]
```diff
@@ -25,27 +38,45 @@ type BearerToken struct {
...
-	endpointURL, err := url.Parse(envVars.PDSControlPlaneURL)
+	var token string
+	currentTestDescription := ginkgo.CurrentGinkgoTestDescription()
+	testName := strings.Split(currentTestDescription.FullTestText, " ")[0]
+	serviceIdFlag := customParams.ReturnServiceIdentityFlag()
+	PDSControlPlaneURL := os.Getenv("CONTROL_PLANE_URL")
+	endpointURL, err := url.Parse(PDSControlPlaneURL)
 	if err != nil {
-		log.Errorf("Unable to access the URL: %s", envVars.PDSControlPlaneURL)
+		log.FailOnError(err, "Unable to connect to the URL- %v", PDSControlPlaneURL)
```
Comment (Madanagopal19) [inline drivers/pds/api/api_helper.go:55]: return like this 

```
if err != nil {
		return nil, fmt.Errorf("Error in getting context for api call: %v\n", err)
	}
```

**Review by Madanagopal19 — APPROVED** [review]
Minor Comment, kindly address it before merging it. Rest looks good to me

**Review by aravindranath14 — APPROVED** [review]
LGTM

**Pull request merged**